### PR TITLE
feat(compression): compress/uncompress functions rewritten.

### PR DESCRIPTION
**Pull request by bouda1**
The goal is to use directly the zlib.

### CODE REVIEW

File: core/src/compression/stream.cc [hunk 1]
```diff
@@ -27,6 +28,111 @@
...
+      logging::error(logging::medium)
+        << "compression: uncompress - Input data is corrupted";
+    return QByteArray();
+  }
+  ulong expected_size = (data[0] << 24) | (data[1] << 16)
+                     | (data[2] <<  8) | data[3];
+  ulong len = qMax(expected_size, 1ul);
+  QByteArray uncompressed_array(len, '\0');
+
+  forever {
```
Comment (ganoze) [inline core/src/compression/stream.cc]: Qt style. Please remove the loop as we have the expected size. Any kind of error from uncompress() can be processed immediately.

Z_OK = return buffer as is.
Z_MEM_ERROR = throw out of memory exception (std::bad_alloc or exceptions::msg).
Z_BUF_ERROR || Z_DATA_ERROR = data corruption error. Let caller process it.

File: core/src/compression/stream.cc [hunk 2]
```diff
@@ -27,6 +28,111 @@
...
+    if (nbytes < 4 || (data[0] != 0 || data[1] != 0
+                   || data[2] != 0 || data[3] != 0))
+      logging::error(logging::medium)
+        << "compression: uncompress - Input data is corrupted";
+    return QByteArray();
+  }
+  ulong expected_size = (data[0] << 24) | (data[1] << 16)
+                     | (data[2] <<  8) | data[3];
+  ulong len = qMax(expected_size, 1ul);
+  QByteArray uncompressed_array(len, '\0');
```
Comment (ganoze) [inline core/src/compression/stream.cc]: Please check expected_size against max_data_size.

File: core/test/compression/stream/write.cc [hunk 3]
```diff
@@ -93,6 +93,25 @@ TEST_F(CompressionStreamWrite, Returns1WithInterleavedRead) {
...
+
+  // When
+  _stream->flush();
+
+  // Then
+  misc::shared_ptr<io::data> d;
+  ASSERT_THROW(_substream->read(d), exceptions::msg);
+}
+
+// When write() is called
```
Comment (ganoze) [inline core/test/compression/stream/write.cc]: Should be removed.

File: core/src/compression/stream.cc [hunk 4]
```diff
@@ -27,6 +28,111 @@
...
+    res = ::compress2(
+        (uchar*)retval.data() + 4,
+        &len,
+        (uchar*)data.constData(),
+        nbytes,
+        compression_level);
+
+    switch (res) {
+      case Z_OK:
+        retval.resize(len + 4);
```
[the remark this answers is not in the document]
Comment (bouda1) [reply core/src/compression/stream.cc]: If we remove this line, we are no more compatible with the previous version.
Moreover, the initial computed _len_ value is a superior bound of the compressed buffer, not its exact length. So, here, _len_ may have changed during the compression, that's why we resize the array.

**Comment by bouda1** [comment]
Done.

File: core/src/compression/stream.cc (hunk 4 above)
[the remark this answers is not in the document]
Comment (ganoze) [reply core/src/compression/stream.cc]: Yeah sure, I forgot that we returned the QByteArray directly. My bad.

File: core/src/compression/stream.cc [hunk 5]
```diff
@@ -27,6 +29,106 @@
...
+        retval[0] = (nbytes >> 24) & 0xff;
+        retval[1] = (nbytes >> 16) & 0xff;
+        retval[2] = (nbytes >> 8) & 0xff;
+        retval[3] = (nbytes & 0xff);
+        break;
+      case Z_MEM_ERROR:
+        throw (exceptions::msg()
+          << "compress: Z_MEM_ERROR: Not enough memory");
```
Comment (ganoze) [inline core/src/compression/stream.cc]: "compression: not enough memory to compress " << nbytes << " bytes"

File: core/src/compression/stream.cc [hunk 6]
```diff
@@ -27,6 +29,106 @@
...
+ *
+ * @param data The data to extract.
+ * @param nbytes The data size in bytes.
+ *
+ * @return the extract data
+ */
+static QByteArray _uncompress(const uchar* data, uLong nbytes) {
+  if (!data) {
+    logging::debug(logging::medium)
+      << "uncompress: data is null";
```
Comment (ganoze) [inline core/src/compression/stream.cc]: "compression: attempting to uncompress null buffer"

File: core/src/compression/stream.cc [hunk 7]
```diff
@@ -27,6 +29,106 @@
...
+    return QByteArray();
+  }
+  if (nbytes <= 4) {
+    if (nbytes < 4 || (data[0] != 0 || data[1] != 0
+                   || data[2] != 0 || data[3] != 0))
+      logging::debug(logging::medium)
+        << "uncompress: data is corrupted";
```
Comment (ganoze) [inline core/src/compression/stream.cc]: "compression: attempting to uncompress data with invalid size"

File: core/src/compression/stream.cc [hunk 8]
```diff
@@ -27,6 +29,106 @@
...
+    return QByteArray();
+  }
+  ulong expected_size = (data[0] << 24) | (data[1] << 16)
+                     | (data[2] <<  8) | data[3];
+  ulong len = qMax(expected_size, 1ul);
+  if (len > (ulong)std::numeric_limits<int>::max())
```
Comment (ganoze) [inline core/src/compression/stream.cc]: Na, there is a variable within compression::stream named max_data_size. If expected_size is greater than max_data_size then the input is corrupted. However do not throw an exception or use a new type like compression::corruption (to be caught by caller).

File: core/src/compression/stream.cc [hunk 9]
```diff
@@ -27,6 +29,106 @@
...
+      (Bytef const*)data + 4, nbytes - 4);
+
+  switch (res) {
+    case Z_OK:
+      if (len != alloc)
+        uncompressed_array.resize(len);
+      return uncompressed_array;
+    case Z_MEM_ERROR:
+      throw (exceptions::msg()
+        << "uncompress: Z_MEM_ERROR: not enough memory.");
```
Comment (ganoze) [inline core/src/compression/stream.cc]: "compression: not enough memory to uncompress " << nbytes " compressed bytes to " << len << " uncompressed bytes";

File: core/src/compression/stream.cc [hunk 10]
```diff
@@ -27,6 +29,106 @@
...
+    case Z_BUF_ERROR:
+    case Z_DATA_ERROR:
+      logging::error(logging::medium)
+        << "uncompress: Z_DATA_ERROR: input data is corrupted";
```
Comment (ganoze) [inline core/src/compression/stream.cc]: Log as debug if wanted. Leave caller handle corruption.

File: core/inc/com/centreon/broker/compression/zlib.hh [hunk 11]
```diff
@@ -0,0 +1,44 @@
...
+** See the License for the specific language governing permissions and
+** limitations under the License.
+**
+** For more information : contact@centreon.com
+*/
+
+#ifndef CCB_COMPRESSION_ZLIB_HH
+#  define CCB_COMPRESSION_ZLIB_HH
+
+#  include <zlib.h>
```
Comment (ganoze) [inline core/inc/com/centreon/broker/compression/zlib.hh]: Please remove inclusion from header.

File: core/inc/com/centreon/broker/compression/zlib.hh [hunk 12]
```diff
@@ -0,0 +1,44 @@
...
+  /**
+   *  @class zlib zlib.hh "com/centreon/broker/compression/zlib.hh"
+   *  @brief Binding around the zlib library.
+   *
+   *  Compress and uncompress data.
+   */
+  class               zlib {
+   public:
+    static QByteArray compress(QByteArray const& data, int compression_level);
+    static QByteArray uncompress(const uchar* data, uLong nbytes);
```
Comment (ganoze) [inline core/inc/com/centreon/broker/compression/zlib.hh]: Use only common types and cast within implementation.

`static QByteArray uncompress(char const* data, unsigned long nbytes);`

File: core/inc/com/centreon/broker/exceptions/corruption.hh [hunk 13]
```diff
@@ -0,0 +1,58 @@
...
+      msg::operator<<(t);
+      return (*this);
+    }
+  };
+}
+
+CCB_END()
+
+#endif // !CCB_EXCEPTIONS_CORRUPTION_HH
+
```
Comment (ganoze) [inline core/inc/com/centreon/broker/exceptions/corruption.hh]: Superfluous line ending.

File: core/src/compression/stream.cc [hunk 14]
```diff
@@ -16,7 +16,10 @@
 ** For more information : contact@centreon.com
 */
 
+#include <limits>
```
Comment (ganoze) [inline core/src/compression/stream.cc]: Useless (I guess).

File: core/src/compression/stream.cc [hunk 15]
```diff
@@ -27,6 +30,7 @@
 using namespace com::centreon::broker;
 using namespace com::centreon::broker::compression;
 
+
```
Comment (ganoze) [inline core/src/compression/stream.cc]: Superfluous line.

File: core/src/compression/stream.cc [hunk 16]
```diff
@@ -143,10 +147,15 @@ bool stream::read(
       // corrupted because the size is greater than the remaining
       // payload size.
       if (_rbuffer.size() >= static_cast<int>(size + sizeof(qint32))) {
-        r->QByteArray::operator=(qUncompress(
-          static_cast<uchar const*>(static_cast<void const*>((
-            _rbuffer.data() + sizeof(qint32)))),
-          size));
+        try {
+          r->QByteArray::operator=(zlib::uncompress(
+            static_cast<uchar const*>(static_cast<void const*>((
```
Comment (ganoze) [inline core/src/compression/stream.cc]: Avoid uchar cast (see zlib::uncompress comments).

File: core/src/compression/zlib.cc [hunk 17]
```diff
@@ -0,0 +1,131 @@
...
+ * @param data the data to compress.
+ * @param compression_level The compression level, by default -1.
+ *
+ * @return The same data compressed.
+ */
+QByteArray zlib::compress(QByteArray const& data, int compression_level) {
+  if (data.isEmpty())
+    return QByteArray(4, '\0');
+
+  uLongf nbytes = (uLongf)data.size();
```
Comment (ganoze) [inline core/src/compression/zlib.cc]: Please use static_cast.

File: core/src/compression/zlib.cc [hunk 18]
```diff
@@ -0,0 +1,131 @@
...
+  if (compression_level < -1 || compression_level > 9)
+    compression_level = -1;
+
+  uLongf len = compressBound(nbytes);
+  QByteArray retval;
+  int res;
+  do {
+    retval.resize(len + 4);
+    res = ::compress2(
+        (Bytef*)retval.data() + 4,
```
Comment (ganoze) [inline core/src/compression/zlib.cc]: Please use 2-spaces indent.
Please use static_cast.

File: core/src/compression/zlib.cc [hunk 19]
```diff
@@ -0,0 +1,131 @@
...
+  if (!data) {
+    logging::debug(logging::medium)
+      << "compression: attempting to uncompress null buffer";
+    return QByteArray();
+  }
+  if (nbytes <= 4) {
+    if (nbytes < 4 || (data[0] != 0 || data[1] != 0
+                   || data[2] != 0 || data[3] != 0))
+    throw (exceptions::corruption()
+        << "compression: attempting to uncompress data with invalid size.");
```
Comment (ganoze) [inline core/src/compression/zlib.cc]: No . at end of log message.

File: core/src/compression/zlib.cc [hunk 20]
```diff
@@ -0,0 +1,131 @@
...
+  }
+  ulong expected_size = (data[0] << 24) | (data[1] << 16)
+                     | (data[2] <<  8) | data[3];
+  ulong len = qMax(expected_size, 1ul);
+  if (len > stream::max_data_size)
+    throw (exceptions::corruption()
+      << "compression: data expected size is too big.");
```
Comment (ganoze) [inline core/src/compression/zlib.cc]: No . at end of log message.

File: core/src/compression/zlib.cc [hunk 21]
```diff
@@ -0,0 +1,131 @@
...
+  QByteArray uncompressed_array(len, '\0');
+
+  ulong alloc = len;
+
+  int res = ::uncompress((Bytef*)uncompressed_array.data(), &len,
```
Comment (ganoze) [inline core/src/compression/zlib.cc]: Please place each argument on separate lines.
Use static_cast.

File: core/src/compression/zlib.cc [hunk 22]
```diff
@@ -0,0 +1,131 @@
...
+      (Bytef const*)data + 4, nbytes - 4);
+
+  switch (res) {
+    case Z_OK:
+      if (len != alloc)
```
Comment (ganoze) [inline core/src/compression/zlib.cc]: Probably useless.

File: core/src/compression/zlib.cc [hunk 23]
```diff
@@ -0,0 +1,131 @@
...
+        uncompressed_array.resize(len);
+      return uncompressed_array;
```
Comment (ganoze) [inline core/src/compression/zlib.cc]: Please break and return uncompressed_array at end of function.

File: core/src/compression/zlib.cc [hunk 24]
```diff
@@ -0,0 +1,131 @@
...
+
+/**
+ * Compression function
+ *
+ * @param data the data to compress.
+ * @param compression_level The compression level, by default -1.
+ *
+ * @return The same data compressed.
+ */
+QByteArray zlib::compress(QByteArray const& data, int compression_level) {
```
Comment (ganoze) [inline core/src/compression/zlib.cc]: Please add unit tests.

File: core/src/compression/zlib.cc [hunk 25]
```diff
@@ -0,0 +1,131 @@
...
+
+/**
+ * Uncompress function
+ *
+ * @param data The data to extract.
+ * @param nbytes The data size in bytes.
+ *
+ * @return the extract data
+ */
+QByteArray zlib::uncompress(const uchar* data, uLong nbytes) {
```
Comment (ganoze) [inline core/src/compression/zlib.cc]: Please add unit tests.

File: core/src/compression/zlib.cc (hunk 24 above)
Comment (bouda1) [reply core/src/compression/zlib.cc]: Done for two cases:
* nominal
* when the buffer to compress is empty

**Pull request merged**